_posts/2024-10-07-comparing-embedders-for-rag.md

```diff
@@ -0,0 +1,264 @@
+    iframe{
+        border: 0px;
+        width: 130em;
+    }
+</style>
+
+<center>
+    <em>Cover image by author.</em>
+</center>
+<br>
+
+### Can we trust LLMs?
+
+When it comes to AI, discerning between hype and reality can sometimes be challenging.
+
+Modern *large language models* (LLM), indeed, fall within the metaphor of **stochastic parrots**.
+
+(***On the Dangers of Stochastic Parrots: Can Language Models Be Too Big?***, by *Emily M. Bender*, *Timnit Gebru*, *Angelina McMillan*, *Shmargaret Shmitchell*)
+
+<center>
+    <img src="{{ site.url }}{{ site.baseurl }}/assets/images/comparing_embedders_for_rag/parrot.png">
+    <br>
+    <em> AI as "stochastic parrot" (Image by author) </em>
+</center>
+
+In short, they are unaware of the language they process, so we cannot fully trust the answers they provide.
+
+So, the answer could be "it depends on our expectations and objectives".
+
+### RAG: "This is the Way"
+
+But what if we wanted to use AI for language-related tasks. ***in a production environment***?
+
+We could follow different strategies reduce hallucinations, such as:
+- **Training** a model on a specific dataset
+- **Fine-tuning** a pretrained model
+- Using **Retrieval Augmented Generation (RAG)**
+
+To achieve our professional goals, RAG stands out as the most suitable option, as it:
+
+1. Avoids <ins>the costs of training/fine-tuning</ins> modeks
+2. Allows for **easier updates of sources** as needed
+
+The idea is to provide an enriched context to the LLM, and then use it only for language synthesis.
+
+In this sense, selecting an effective embedder plays a crucial role in the retrieval process.
+
+In fact, the generation of embeddings is the basis of the process of retrieval.
+
+**This article** steps into a possible approach to **evaluate various embedders** for a specific language.
+
+#### STEP 1: Creating samples
+
+I created samples with the following structure:
+
+```json
+{
+    "Topic": [
+        "topic-related_sentence_1",
+        "topic-related_sentence_2",
+        // ...
+    ]
+}
+```
+
+  Example:
+```json
+    {
+        "Tecnologia": [
+        "Il computer è uno strumento.",
+        "I telefoni moderni sono versatili.",
+        "La realtà virtuale è immersiva."
+    ],
+    "Cucina": [
+        "Gli chef creano piatti deliziosi.",
+        "I dolci sono molto apprezzati.",
+        "Il forno è molto utile."
+    ]
+    }
+```
+
+#### STEP 2: Downloading models
+
+After conducting some research, I compiled a list of effective embedders for Italian:
+
+- [Anita](https://huggingface.co/DeepMount00/Anita)
+- [cross_encoder_italian_bert_stsb](https://huggingface.co/nickprock/cross-encoder-italian-bert-stsb)
+- [mmarco_bert_base_italian_uncased](https://huggingface.co/nickprock/mmarco-bert-base-italian-uncased)
+- [multi_e5](https://huggingface.co/intfloat/multilingual-e5-large)
+- [other models ...]
+
+Subsequently, I used ***transformer*** library to download and tokenize models, iterating over my list.
```

Comment:
missing link to the library and avoid *iterating over my list*. What *tokenize a model* means?